src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -3,349 +3,515 @@
+        self.state = ClientState.SMS_AUTH_REQUESTED
+
+    def sms_code(self, code: str) -> None:
+        """Sends the SMS auth code, completing authentication.
+
+        :raises InvalidSmsAuthError:
+        """
+        assert self.state is ClientState.SMS_AUTH_REQUESTED
+        assert self._trusted_phone is not None
+        url = "https://idmsa.apple.com/appleauth/auth/verify/phone/securitycode"
+        logger.debug("PUT {url}")
+        response = self.session.post(
+            url,
+            json={
+                "securityCode": {"code": code},
+                "phoneNumber": {"id": self._trusted_phone.id},
+                "mode": self._trusted_phone.push_mode,
+            },
+            headers={
+                "scnt": self._scnt,
+                "X-Apple-Id-Session-Id": self.session_id,
+                "Accept": "application/json",
+                "X-Apple-Widget-Key": self.service_key,
+            },
+        )
+        if response.status_code != HTTPStatus.OK:
+            # TODO: Make invalid code distinguishable from generic error.
+            raise InvalidAuthCodeError
+        else:
+            self.state = ClientState.AUTHENTICATED
+
+        # If we want more info about the failure extract failure info from the response body
+        # response is a JSON with the following interesting fields:
+        # (Extracted from a login attempt with wrong code)
+        # 'restrictedAccount': False,
+        # 'securityCode': {'code': '123123',
+        #                  'securityCodeCooldown': False,
+        #                  'securityCodeLocked': False,
+        #                  'tooManyCodesSent': False,
+        #                  'tooManyCodesValidated': False},
+        # 'serviceErrors': [{'code': '-21669',
+        #                    'message': 'Incorrect verification code.',
+        #                    'suppressDismissal': False,
+        #                    'title': 'Incorrect Verification Code'}],
+
+    def session_cookie(self) -> str:
+        """Extracts the session cookies.
+
+        This can be used to create a new client without having to re-authenticate, if the
+        session is still valid.
+
+        :returns: the session cookie.
+        :raises: an exception if the client is not yet authenticated and has no session
+           cookie.
+        """
+        assert self.state is ClientState.AUTHENTICATED
+        return self.session.cookies.get(SESSION_COOKIE_NAME)  # type: ignore
+
+    def load_session_cookie(self, cookie: str) -> None:
+        """Loads the itunes session cookie in the client's session.
+
+        This will invalidate the client's authentication state and reset it.  It will
+        trigger a request to check a cookie is valid and use this to set the state
+        correctly.
+
+        If the iTunes session is still valid the user will be logged in inside the session.
+        """
+        self.session.cookies.set(self.SESSION_COOKIE_NAME, cookie)  # type: ignore
```

Comment:
similarly here the class docs state that this should only be used when the client is already authenticated, but it doesn't look like there isn't any checking on the state done here.

`from_session_cookie`, if i understand correctly will invoke this while `state == ClientState.NEW` and contradicts the class docs a little.

Reply:
the class docs don't mention this i think.  though maybe we can just remove this method and only expose `.from_session_cookie`.  This is a bit weird.